### PR TITLE
GPL-741 - Add function to report from a particular date

### LINKED ISSUE #167: GPL-741 Should we change the +ves report file format
**User story**
GPL-741 | As a dev manage LIMS (Rich L) we may need to change the format of the +ves report to avoid truncating actionable data

**Who are the primary contacts for this story**
Rich L
Rich C
PAM
Ian J

**Options**

1. Remove samples form the report that are over 6 weeks old (Preferred by operations discussed stand up of 13th Nov)
2. Change the file to a text file format
3. Produce more than one file 

**Acceptance criteria**
To be considered successful the solution must allow:
- [x] change file to sliding 6 week date window

**Dependencies**
Limit of MS files 1,048,576 rows (1048k) : https://support.microsoft.com/en-us/office/excel-specifications-and-limits-1672b34d-7043-467e-8e27-269d656771c3 

**Additional context**
Need for output may be superseded by New Beckman robotics package WIP Nov
+ves report at ---k rows on 20th Nov (--k/day)
+ves report at 920k rows on 19th Nov (10k/day)
+ves report at 910k rows on 18th Nov (18k/day)
+ves report at 892k rows on 17th Nov (36k/day)
+ves report at 856k rows on 16th Nov (37k/day)
+ves report at 819k rows on 13th Nov (18k/day)
+ves report at 602k rows on 1st Nov (14.7k/day)
+ves report at 336k rows on 13th Oct

### PR DESCRIPTION
**Pull request by pjvv**
Closes #167 

Changes proposed in this pull request:

* add function to get window start date
* use mongo aggregate function

### CODE REVIEW

File: lighthouse/helpers/plates.py [hunk 1]
```diff
@@ -134,10 +133,27 @@ def get_samples(plate_barcode: str) -> Optional[List[Dict[str, Any]]]:
...
+    """
+    samples_collection = app.data.driver.db.samples
+
+    # The pipeline defines stages which execute in sequence
+    pipeline = [
+        # 1. First run the positive match stage
+        STAGE_MATCH_POSITIVE,
+        # 2. We are only interested in the samples for a particular plate
+        {"$match": {FIELD_PLATE_BARCODE: plate_barcode}},
+    ]
```
Comment (emrojo) [inline lighthouse/helpers/plates.py:152]: should this be the other way? first match the barcode and then match the positive? it should be quicker:

https://docs.mongodb.com/manual/core/aggregation-pipeline-optimization/

File: lighthouse/constants.py [hunk 2]
```diff
@@ -37,19 +36,35 @@
...
+            {
+                "$or": [
+                    {FIELD_CH1_CQ: {"$lte": CT_VALUE_LIMIT}},
+                    {FIELD_CH2_CQ: {"$lte": CT_VALUE_LIMIT}},
+                    {FIELD_CH3_CQ: {"$lte": CT_VALUE_LIMIT}},
+                ],
+            },
+        ],
+        # 4. We are only interested in documents which have a valid date
+        FIELD_DATE_TESTED: {"$exists": True, "$nin": [None, ""]},
```
Comment (emrojo) [inline lighthouse/constants.py:68]: Just to let you know we also have the sample creation date  (Field ```created_at```) That is the date when the sample was first recorded in the Mongo database. What to use depends in what the SSRs are more interested, if they want the recent samples by testing date, or the recent samples by how we are receiving them (created_at).

File: lighthouse/helpers/reports.py [hunk 3]
```diff
@@ -241,25 +238,85 @@ def get_cherrypicked_samples(root_sample_ids, plate_barcodes, chunk_size=50000):
...
+                    "$dateFromString": {
+                        "dateString": f"${DATE_EXTRACTED}",
+                        "format": "%Y-%m-%d %H:%M:%S",
+                        "timezone": "UTC",
+                        "onError": {
+                            "$dateFromString": {
+                                "dateString": f"${DATE_EXTRACTED}",
+                                "format": "%d/%m/%Y %H:%M",
+                                "timezone": "UTC",
+                                "onError": DATE_ERROR,
```
Comment (emrojo) [inline lighthouse/helpers/reports.py:298]: I'm thinking with this, we will never see in the report any sample that has a testing date not matching any of these 2 conditions. Do these 2 conditions match all testing dates in database? If that is the case, then is fine for the moment and we should replace this in future with a migration to use store this testing date as Date in that field (so the error on format gets captured before); but if we can have new data from plate maps that doesn't match these 2 conditions, then we could miss new samples in the report. Wouldn't it be easier in that case to use 'created_at' instead?

File: lighthouse/helpers/plates.py (hunk 1 above)
Comment (pjvv) [reply lighthouse/helpers/plates.py:152]: Yes, good spot!

File: lighthouse/constants.py (hunk 2 above)
Comment (pjvv) [reply lighthouse/constants.py:68]: Rich (SSR) said they would like to use the "Date Tested" date.

File: lighthouse/helpers/reports.py (hunk 3 above)
Comment (pjvv) [reply lighthouse/helpers/reports.py:298]: Yes, I need to create an issue in the crawler for this.

**Review by emrojo — DISMISSED** [review]

**Pull request merged**